service/src/main/java/uk/gov/hmcts/reform/fpl/request/CacheAwareRequestData.java

```diff
@@ -0,0 +1,31 @@
+package uk.gov.hmcts.reform.fpl.request;
+
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.context.annotation.Primary;
+import org.springframework.stereotype.Service;
+
+import javax.servlet.http.HttpServletRequest;
+
+@Primary
+@Service
+public class CacheAwareRequestData implements RequestData {
+
+    private final HttpServletRequest httpServletRequest;
+
+    @Autowired
+    public CacheAwareRequestData(HttpServletRequest httpServletRequest) {
+        this.httpServletRequest = httpServletRequest;
```

Comment:
Does this servlet request still have a possibility of getting cleaned up before we have used it?

Reply:
If RequestData is accessed from async task then AsyncConfiguration guarantees that there will be cached version of RequestData and httpRequest wont be consulted.
If Request data is accessed from standard "request"  thread then it is guaranteed by spring request scope that httpServletRequest is valid.
The question is about when exactly cached version is created, and it created in request thread.

 @Override
        public Runnable decorate(@Nonnull Runnable task) {
            //request thread where we create copy of request data
            SimpleRequestData requestData = new SimpleRequestData(context.getBean(RequestData.class));

           //everything in lambda is executed in new task thread
            return () -> {
                RequestDataCache.add(requestData);
                try {
                    task.run();
                } finally {
                    RequestDataCache.remove();
                }
            };
        }

Reply:
To test it i have setup code like this:
    @PostMapping("/submitted")
    public void handleSubmittedEvent(@RequestBody CallbackRequest callbackRequest) {
        System.out.println("Invoking async action " + Thread.currentThread().getName());
        localAuthorityUserService.grantUserAccessWithCaseRole(caseId, caseLocalAuthority);
    }

in AsyncConfiguration
 @Override
        public Runnable decorate(@Nonnull Runnable task) {
            System.out.println("Copying request data " + Thread.currentThread().getName());
            SimpleRequestData requestData = new SimpleRequestData(context.getBean(RequestData.class));

            return () -> {
                System.out.println("Setting thread local"+Thread.currentThread().getName());
                RequestDataCache.add(requestData);
                try {
                    task.run();
                } finally {
                    RequestDataCache.remove();
                }
            };
        }

output is: 
2020-06-24T17:32:15.734 INFO  [http-nio-4000-exec-1] u.g.h.r.logging.filters.RequestStatusLoggingFilter Request POST /callback/case-initiation/about-to-submit processed in 242ms
Invoking async action http-nio-4000-exec-2
Copying request data http-nio-4000-exec-2
Setting thread localgetAsyncExecutor-1

which proves that copy of request data is done in request thread where correct httpRequest is guranteed by spring

Reply:
Sounds good to me.
On a slight side note if you care, you can change the name from `localgetAsyncExecutor` to something else in the `@Bean(name="something")`